### ISSUE TITLE
New linter to simplify negatives in conditional expressions?

### ISSUE THREAD

**Issue by IndrajeetPatil** [r-lib/lintr#1831]
## Preamble

There ain't nobody that doesn't struggle with negatives, and it would be nice to have a linter that suggests a way to simplify boolean tests with negatives.

Here are a couple of ways to do this in a single linter.

## Re-order `if`/`else`

Actual

```r
if (!normal_case) {
  ...1
} else {
  ...2
}
```

Suggested

```r
if (normal_case) {
  ...2
} else {
  ...1
}
```

## Applying De Morgan's laws

Actual = Suggested

- (not A) or (not B) = not (A and B)

```r
if (!has_fuel || !has_mirrors) {
  can_be_rented <- FALSE
}

if (!(has_fuel && has_mirrors)) {
  can_be_rented <- FALSE
}
```

- (not A) and (not B) = not (A or B)

```r
if (!is_tank_empty && !is_mirror_broken) {
  can_be_rented <- TRUE
}

if (!(is_tank_empty || is_mirror_broken)) {
  can_be_rented <- TRUE
}

```

WDYT? Any other ways to simplify these?

**Comment by AshesITR** [comment]
Honestly, I prefer `!has_fuel || !has_mirrors` because parentheses use additional brainpower.

Also, control flow should be ordered in such a way that the surrounding code reads naturally. Therefor, it depends on context whether `if (!cond)` or `if (cond)` is more readable.

**Comment by MichaelChirico** [comment]
this is a Google linter IINM :)

**Comment by IndrajeetPatil** [comment]
> Honestly, I prefer `!has_fuel || !has_mirrors` because parentheses use additional brainpower.

Fair enough. But my preferred approach here would be to extract the conditional expressions to a new variable so that the parentheses can be avoided. Additionally, this introduces a reusable abstraction with an informative name. 

```r
# actual
if (!has_fuel || !has_mirrors) {
  can_be_rented <- FALSE
}

# refactored
is_car_in_good_condition <- has_fuel && has_mirrors
if (!is_car_in_good_condition) {
  can_be_rented <- FALSE
}
```

**Comment by MichaelChirico** [comment]
> I sense there is no consensus about whether the second part should be supported.

I think consensus is more important for setting defaults -- it seems like a reasonable option to add, though I wouldn't enable it by default.

**Comment by IndrajeetPatil** [comment]
Sounds good to me.

Maybe we can call this parameter `enforce_de_morgan_laws` and set it `FALSE` by default?

**Comment by AshesITR** [comment]
`use_de_morgans_laws = FALSE` is more in line with our other parameter names. Also note the genitive "s".
